File: backend/app/routers/donors.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session

from ..database import get_db
from ..deps import require_any_permission
from ..models import Donation, Donor
from ..schemas import DonorGiftOut, DonorOut

router = APIRouter(
    prefix="/api/donors",
    tags=["donors"],
    dependencies=[Depends(require_any_permission("donors", "pledge-campaign-pledges"))],
)


@router.get("", response_model=list[DonorOut])
def list_donors(db: Session = Depends(get_db)) -> list[Donor]:
    """The persistent Giving App donor list - lives under Config as "Giving
    App - Donors" and doubles as the lookup for the Pledges page's donor
    picker."""
    return list(db.scalars(select(Donor).order_by(Donor.last_name, Donor.first_name)))


@router.get("/{donor_id}/gifts", response_model=list[DonorGiftOut])
def donor_gifts(donor_id: str, db: Session = Depends(get_db)) -> list[Donation]:
    """Every gift this donor has given, across every fund - not scoped to
    one campaign, for the Donors page's click-to-expand detail popup."""
    if db.get(Donor, donor_id) is None:
        raise HTTPException(404, "Donor not found.")
    return list(
        db.scalars(
            select(Donation)
            .where(Donation.donor_id == donor_id)
            .order_by(Donation.received_date.desc().nulls_last())
        )
    )

File: frontend/src/api/donors.ts
// The persistent Giving App donor list - Config > "Giving App - Donors",
// and the lookup behind the Pledges page's donor picker.
import { BASE, authHeaders, j } from "./client";

export interface Donor {
  donor_id: string;
  donor_number: string;
  first_name: string;
  last_name: string;
  email: string;
  phone_number: string;
  city: string;
  state: string;
  zip_code: string;
  joint_giver_id: string;
  joint_giver_first_name: string;
  joint_giver_last_name: string;
  first_donated: string | null;
  donation_count: number;
  total_given: number;
  source_file_name: string;
  source_file_link: string;
}

export interface DonorGift {
  id: number;
  fund: string;
  received_date: string | null;
  amount: number;
  net_amount: number;
  method: string;
  source_file_name: string;
  source_file_link: string;
}

export const donorsApi = {
  list: () => fetch(`${BASE}/api/donors`, { headers: authHeaders() }).then(j<Donor[]>),

  /** Every gift this donor has given, across every fund - not scoped to
   * one campaign, for the Donors page's click-to-expand detail popup. */
  gifts: (donorId: string) =>
    fetch(`${BASE}/api/donors/${encodeURIComponent(donorId)}/gifts`, { headers: authHeaders() }).then(
      j<DonorGift[]>
    ),
};
